recon.py

```diff
@@ -0,0 +1,329 @@
+    Cospeciation is an event where two children of the parasite vertex
+    maps to the two children of the host vertex. Cospeciation event has
+    two child MappingNode named left and right.
+    """
+
+    @property
+    def event_type(self) -> EventType.COSPECIATION:
+        return EventType.COSPECIATION
+
+class Duplication(TwoChildrenEvent):
+    """
+    Duplication is an event where two children of the parasite vertex
+    maps to the same host vertex as their parent. Cospeciation event has
+    two child MappingNode named left and right.
+    """
+
+    @property
+    def event_type(self) -> EventType:
+        return EventType.DUPLICATION
+
+class Transfer(TwoChildrenEvent):
+    """
+    Transfer is an event where one child of the parasite vertex
+    maps to the same host vertex as its parent, while the other child maps 
+    to a host vertex not ancestrally related to the first one. Transfer event 
+    has two child MappingNode named left and right. The right child is always
+    the child that gets transferred.
+    """
+
+    @property
+    def event_type(self) -> EventType:
+        return EventType.TRANSFER
+
+class Loss(Event):
+    """
+    Loss is an event where the parasite vertex goes down to only one of 
+    the lineage of the host vertex when the host vertex speciates. Loss event
+    has one child MappingNode called child. The parasite of the child MappingNode
+    (self.child.parasite) is the same as the parasite of the MappingNode that
+    incurs the lost.
+    """
+
+    def __init__(self, child: MappingNode, freq=None):
+        """
+        Creates a Loss event. The lost event has one child MappingNode
+        called child.
+        """
+        super().__init__(freq)
+        self._child = child
+
+    @property
+    def child(self) -> MappingNode:
+        return self._child
+    
+    def __eq__(self, other):
+        return type(self) == type(other) and self._child == other.child
+    
+    def __hash__(self):
+        return hash(self._child)
+    
+    def __repr__(self):
+        return "%s(%s)" % (type(self).__name__, self._child)
+
+    @property
+    def event_type(self) -> EventType:
+        return EventType.LOSS
+
+class TipTip(Event):
+    """
+    TipTip (tip mapping) is an event where both the parasite and the host are 
+    leaves. Biologically, this means the parasite and the host lives together
+    at present time. The TipTip event is the sink of the reconciliation graph and
+    has no children.
+    """
+
+    def __init__(self, freq: float = None):
+        """
+        Creates a TipTip (tip to tip mapping) event. The TipTip event has 
+        no children mapping nodes.
+        """
+        super().__init__(freq)
+
+    def __repr__(self):
+        return "%s()" % type(self).__name__
+
+    @property
+    def event_type(self) -> EventType:
+        return EventType.TIPTIP
+
+class Reconciliation:
+    """
+    Reconciliation is a tree that represents an edge-to-edge mapping from 
```

Comment:
Reconciliation represents a single reconciliation in a reconciliation graph (ReconGraph).  The reconciliation is a subgraph of a the reconciliation graph beginning at a source mapping node of the reconciliation graph and terminating at all of the sink nodes of the reconciliation graph.  Each MappingNode...